src/tsc.cpp

```diff
@@ -7,43 +7,47 @@
 #include "tsc.h"
 #include "vmEntry.h"
 
-
 bool TSC::_initialized = false;
 bool TSC::_available = false;
 bool TSC::_enabled = false;
 u64 TSC::_offset = 0;
 u64 TSC::_frequency = NANOTIME_FREQ;
 
-
 void TSC::enable(Clock clock) {
-    if (!TSC_SUPPORTED || clock == CLK_MONOTONIC || !VM::loaded()) {
+    if (!TSC_SUPPORTED || clock == CLK_MONOTONIC) {
         _enabled = false;
         return;
     }
 
     if (!_initialized) {
-        JNIEnv* env = VM::jni();
-
-        jfieldID jvm;
-        jmethodID getTicksFrequency, counterTime;
-        jclass cls = env->FindClass("jdk/jfr/internal/JVM");
-        if (cls != NULL
-                && ((jvm = env->GetStaticFieldID(cls, "jvm", "Ljdk/jfr/internal/JVM;")) != NULL)
-                && ((getTicksFrequency = env->GetMethodID(cls, "getTicksFrequency", "()J")) != NULL)
-                && ((counterTime = env->GetStaticMethodID(cls, "counterTime", "()J")) != NULL)) {
+        if (VM::loaded()) {
+            JNIEnv* env = VM::jni();
+
+            jfieldID jvm;
+            jmethodID getTicksFrequency, counterTime;
+            jclass cls = env->FindClass("jdk/jfr/internal/JVM");
+            if (cls != NULL
+                    && ((jvm = env->GetStaticFieldID(cls, "jvm", "Ljdk/jfr/internal/JVM;")) != NULL)
+                    && ((getTicksFrequency = env->GetMethodID(cls, "getTicksFrequency", "()J")) != NULL)
+                    && ((counterTime = env->GetStaticMethodID(cls, "counterTime", "()J")) != NULL)) {
+                u64 frequency = env->CallLongMethod(env->GetStaticObjectField(cls, jvm), getTicksFrequency);
+                if (frequency > NANOTIME_FREQ) {
+                    // Default 1GHz frequency might mean that rdtsc is not available
+                    u64 jvm_ticks = env->CallStaticLongMethod(cls, counterTime);
+                    _offset = rdtsc() - jvm_ticks;
+                    _frequency = frequency;
+                    _available = true;
+                }
+            }
 
-            u64 frequency = env->CallLongMethod(env->GetStaticObjectField(cls, jvm), getTicksFrequency);
-            if (frequency > NANOTIME_FREQ) {
-                // Default 1GHz frequency might mean that rdtsc is not available
-                u64 jvm_ticks = env->CallStaticLongMethod(cls, counterTime);
-                _offset = rdtsc() - jvm_ticks;
-                _frequency = frequency;
+            env->ExceptionClear();
+            _initialized = true;
```

Comment:
Why is it set to `true` only in the VM branch? The initialization code is supposed to run once in non-Java case too, isn't it?

Reply:
I didn't understand the code logic. fixed that.

Reply:
ed: when I wrote the code I tried to follow the pre-existing logic, which didn't set `_initialized` if the JVM was not loaded. I think that it is better to do it this way.